### ISSUE TITLE
domish.Element.__str__ does not always return a string

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3579]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)| @washort reported|
|-|-|
|Trac ID|trac#3579|
|Type|defect|
|Created|2008-12-15 18:16:32Z|

The type check in `__str__` allows unicode objects to be returned, as seen below:
```
>>> from twisted.words.xish.domish import Element
>>> e = Element((None, "foo"))
>>> e.addContent(u'\N{SNOWMAN}')
u'\u2603'
>>> str(e)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2603' in position 0: ordinal not in range(128)
```

<details><summary>Searchable metadata</summary>

```
trac-id__3579 3579
type__defect defect
reporter__washort washort
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1229364992000000 1229364992000000
changetime__1466078289518541 1466078289518541
version__None None
owner__ralphm ralphm
cc__exarkun cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @ralphm|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm set status to `assigned`|
|-|-|

This is one of those things that are icky about `domish.Element`. The `__str__` has been defined like this from the start:

```python
    def __str__(self):
        """ Retrieve the first CData (content) node
        """
        for n in self.children:
            if isinstance(n, types.StringTypes): return n
        return ""
```

I agree this is horribly wrong, but I don't know to what extend people use `str(element)` to retrieve it's (unicode) string representation. I think I use `unicode(element)` everywhere nowadays.

If we change this behaviour, what should `__str__` return? A UTF-8 encoded string of the character data?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Why is the API for retrieving the first CDATA child of an `Element` `__str__`?  That's insane.  **Insane**.  In_sane_.

Why is there even an API for retrieving the first CDATA child of an `Element`?  What's the use-case for this behavior?

```python
>>> from twisted.words.xish.domish import Element
>>> e = Element((None, "foo"))
>>> e.addChild('foo')
'foo'
>>> e.addChild('bar')
'bar'
>>> str(e)
'foo'
>>>
```

`'foo'`?  Seriously?

Even if that's somehow useful, `__str__` is a bad way to expose it.  A real method is the right API.  Ditto for `__unicode__`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

I wholeheartedly agree about the inappropriateness of `__str__` as the name for this API.  The repetition of "foo" makes your previous comment about its insanity a bit confusing, though.  To be clear for other potential readers of this ticket:

```
#!py
>>> from twisted.words.xish.domish import Element
>>> e = Element((None, 'foo'))
>>> e.addChild('bar')
'bar'
>>> e.addChild('baz')
'baz'
>>> str(e)
'bar'
```
And, to stress how this won't make any sense if you're trying to debug it:
```
#!py
>>> e.toXml()
u'<foo>barbaz</foo>'
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm commented|
|-|-|

First of all: I agree that this interface is horrible.

However, exarkun's example is flawed in that he uses `addChild` instead of `addContent`, although I notice that the latter is not in `IElement`. This method appends the given text to the CDATA if the current last child of the element is also a text node:

```python
>>> from twisted.words.xish.domish import Element
>>> e = Element((None, "foo"))
>>> e.addContent('foo')
'foo'
>>> e.addContent('bar')
'foobar'
>>> str(e)
'foobar'
```

The use case for retrieving the first text child node of an element is when the typical use of the XML is that an element either has only CDATA, or only child elements. This is very common in Jabber. It doesn't make much sense for mixed-content documents. An example XML stanza:

```
#!xml
<presence from='user@example.com/Home' to='user@example.org'>
  <show>away</show>
  <status>Not here right now</status>
  <priority>1</priority>
</presence>
```

In that example, retrieving the values for `show`, `status` and `priority` can be, currently, done with `unicode(presence.show)`.

There is a lot of code that depends on this (and other questionable API design features). I don't think we can change much about it other than providing an alternate, more standard API. The requirements listed in [#3201](https://github.com/twisted/twisted/issues/3201) would still need to be addressed, though.

Suggestions welcome.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|


Replying to [ralphm](#note_5):
> First of all: I agree that this interface is horrible.

Your earlier comment implied this, but thanks for making it clear.  Good to know that nobody's completely crazy here :).

> However, exarkun's example is flawed in that he uses `addChild` instead of `addContent`,

An interesting digression, but given the similarity of the resulting structures, perhaps this is simply a bug in `addChild`?  (and certainly in `IElement`).

> The use case for retrieving the first text child node of an element is when the typical use of the XML is that an element either has only CDATA, or only child elements. This is very common in Jabber. It doesn't make much sense for mixed-content documents. An example XML stanza:

Is this stanza meant to show that `<status>`, `<priority>` et. al. may _only_ contain text?  i.e. you can't have a boldfaced word inside a status message?  That seems ... limiting.

> In that example, retrieving the values for `show`, `status` and `priority` can be, currently, done with `unicode(presence.show)`.

I think you mean, if `show`, `status` and `priority` all _happen_ to be bytestrings in your particular `Element` and all _happen_ to be encoded in your current default encoding?

> There is a lot of code that depends on this (and other questionable API design features). I don't think we can change much about it other than providing an alternate, more standard API.

Absolutely.  All the [usual stuff about incompatible changes](http://twistedmatrix.com/trac/wiki/CompatibilityPolicy) applies.

> The requirements listed in [#3201](https://github.com/twisted/twisted/issues/3201) would still need to be addressed, though.

There are similar methods, [like getNodeText in domhelpers](http://twistedmatrix.com/documents/8.1.0/api/twisted.web.domhelpers.html#getNodeText); maybe `Element` should implement the appropriate APIs so that it can just be one of those.  One of the reasons that [#3201](https://github.com/twisted/twisted/issues/3201) was filed in the first place is that it's annoying to have to re-implement every XML helper API for domish/(micro/mini)dom/lxml/etree (etc etc); a step in the right direction would be at least factoring out the convenience stuff.

> Suggestions welcome.

To resolve _this_ issue, `__str__` should be deprecated.  It's a bit weird to deprecate `__str__` (or `__unicode__`), but then, it's a bit weird to use it this way in the first place.

There should be some documentation, and a new name for the desired convenience behavior.  If `domhelpers` is an unsuitable place, maybe `IElement.simpleTextContent`?  I'm still not _quite_ clear enough on the use-case here to give it an optimal name.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm commented|
|-|-|

Replying to [glyph](#note_6):
>
> Replying to [ralphm](#note_5):
>
> [..]
>
> > However, exarkun's example is flawed in that he uses `addChild` instead of `addContent`,
>
> An interesting digression, but given the similarity of the resulting structures, perhaps this is simply a bug in `addChild`?  (and certainly in `IElement`).
> > The use case for retrieving the first text child node of an element is when the typical use of the XML is that an element either has only CDATA, or only child elements. This is very common in Jabber. It doesn't make much sense for mixed-content documents. An example XML stanza:
>
> Is this stanza meant to show that `<status>`, `<priority>` et. al. may _only_ contain text?  i.e. you can't have a boldfaced word inside a status message?  That seems ... limiting.

I'd say the actual flaw is not in `addChild`, but rather in retrieving the node text. Whatever the name of the method, it should return the string of concatenated child text nodes values. In XPath you'd write that as: `string(show)` if the context node is the presence stanza.

Indeed, `<status/>`, `<priority/>' may only contain text nodes as childs. Jabber protocols rarely use so-called mixed-content XML (where a particular node may have both text nodes and element nodes as direct childs).

To expand a bit on this to have a better idea of Jabber protocols, we have specified [http://www.xmpp.org/extensions/xep-0071.html XHTML-IM (XEP-0071)] for simple markup in text messages. The idea is that there's always a plain-text version in the normal `<body/>` child of a message, and optionally an XHTML-IM one that is a richer representation.

> > In that example, retrieving the values for `show`, `status` and `priority` can be, currently, done with `unicode(presence.show)`.
>
> I think you mean, if `show`, `status` and `priority` all ''happen'' to be bytestrings in your particular `Element` and all ''happen'' to be encoded in your current default encoding?

Yeah, the assumption here is that the Element was created by the parsing process of incoming XML, or that the text was site with `addContent` or using `addElement('foo', content='my text')`. Unfortunately there are several places that accept both `str` and `unicode` as argument and they are both used as-is internally, too. Element should really only deal with `unicode`.

> [..]
>
> > The requirements listed in #3201 would still need to be addressed, though.
>
> There are similar methods, [http://twistedmatrix.com/documents/8.1.0/api/twisted.web.domhelpers.html#getNodeText like getNodeText in domhelpers]; maybe `Element` should implement the appropriate APIs so that it can just be one of those.  One of the reasons that #3201 was filed in the first place is that it's annoying to have to re-implement every XML helper API for domish/(micro/mini)dom/lxml/etree (etc etc); a step in the right direction would be at least factoring out the convenience stuff.

Unfortunately, that module doesn't have much documentation. Is it modelled after some other DOM in terms of method names and arguments?

> > Suggestions welcome.
>
> To resolve ''this'' issue, `__str__` should be deprecated.  It's a bit weird to deprecate `__str__` (or `__unicode__`), but then, it's a bit weird to use it this way in the first place.

Yeah, exarkun and I discussed this for a bit yesterday. The same would probably need to be done for accessing XML children through python attributes on the Element. That has its own set of issues.

I don't think accessing attributes through subscription is particularly bad, but its implementation could be better. For example, this is pretty bad:

{{{
#!python
>>> from twisted.words.xish import domish
>>> foo = domish.Element((None, 'foo'))
>>> foo['bar'] = 'baz'
>>> foo.toXml()
u"<foo bar='baz'/>"
>>> 'bar' in foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.5/site-packages/twisted/words/xish/domish.py", line 428, in __getitem__
    return self.attributes[self._dqa(key)]
KeyError: 0
}}}

> There should be some documentation, and a new name for the desired convenience behavior.  If `domhelpers` is an unsuitable place, maybe `IElement.simpleTextContent`?  I'm still not ''quite'' clear enough on the use-case here to give it an optimal name.

I think the [http://twistedmatrix.com/documents/8.1.0/api/twisted.web.domhelpers.html#gatherTextNodes `gatherTextNodes`] method does what we'd want here. I'm a bit sceptic about its `dounescape` parameter, though. In XML you just have CDATA and that would be represented as `unicode` without any escaping or entity references. On the other hand, elementtree has the `text` attribute on its Element instances. It holds all text before the first child element.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_8"></a>|@ralphm commented|
|-|-|

[#4232](https://github.com/twisted/twisted/issues/4232) calls for changing this method as well. They are probably not duplicates, though.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm commented|
|-|-|

In changeset 98bcc7ff0bc2c44fbcc24b2c360fe0ff179b965c
```
#!CommitTicketReference repository="" revision="98bcc7ff0bc2c44fbcc24b2c360fe0ff179b965c"
Fix reliance on using `str` for extracting an element's CDATA.

Due to an unfortunate design choice, the typical interface to extract
the CDATA inside a `domish.Element` is using `str` on the element and is
implemented by `__str__` returning the first string-like object in its
list of children, regardless of being bytes or unicode strings.
See #3579 for details.

For now, use `unicode` instead, even though the actual implementation is
in `__str__`. This bug became more obvious because of the change to
`io.StringIO` for the Python 3 port.
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

https://gist.github.com/hawkowl/bbc722cca76209fd9c67a2f4709f77cd might provide some help with this; on 2 it produces:

```
str
str
Unicode
```

and on 3:

```
str
b'bytes'
str
```

So, the correct thing seems to be that you implement a ```__unicode__``` on Py2 (but not py3) and a ```__bytes__``` on Py3 (but not py2), and have ```___str___``` be the bytes impl on 2 and the unicode impl on 3.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_12"></a>|@ralphm set owner to @ralphm<br>@ralphm set status to `new`|
|-|-|

Working on porting xish over to Python 3 ([#8337](https://github.com/twisted/twisted/issues/8337)), this came up again. Because `__str__` is meant to return `unicode` in Python 3, the original implementation is correct for it. However, calling `bytes()` on it would yield a cryptic message like this (because `__bytes__` is not defined):

```
Traceback (most recent call last):
  File "/home/ralphm/src/Twisted/twisted/words/test/test_domish.py", line 136, in test_characterDataBytes
    text = bytes(element)
  File "/home/ralphm/src/Twisted/twisted/words/xish/domish.py", line 450, in __getitem__
    return self.attributes[self._dqa(key)]
builtins.KeyError: 0
```

I did some research on actual usage of `str()` on elements (vs. the more correct `unicode()`) and found that in almost all cases, the result of this call is used to extract identifiers or fixed (ASCII-only) values from elements. This is consistent with how XMPP protocols are designed. The reason it works in Python 2 is that string comparisons between `str` and `unicode` objects yield true if the `str` can be coerced into a `unicode` string. I.e. only contains ASCII.

In Python 3 this doesn't work anymore, of course. However I do feel that `__bytes__` should return something more useful.

In my current working copy I have renamed `__str__` to `__unicode__`, created a `__bytes__` that returns `unicode(self).encode('utf-8')`, and set `__str__` to `__unicode__` on Py3 and to `__bytes__` on Py2.

FWIW I contemplated encoding to `ascii`. I am not sure if this is better.

Opinions?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

That looks about right. The spec says UTF-8 (https://xmpp.org/rfcs/rfc3920.html#xml-encoding).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_14"></a>|@ralphm commented|
|-|-|

In changeset 124628b4f993daf8fb2a1a026ed2dfad9a769da2
```
#!CommitTicketReference repository="" revision="124628b4f993daf8fb2a1a026ed2dfad9a769da2"
Make bytes() return the UTF-8, unicode() the unicode of the CDATA.

This addresses #3579.
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_15"></a>|@ralphm commented|
|-|-|

In changeset 0ec7c65fa6f7592492c4ae73536408caf25c29b3
```
#!CommitTicketReference repository="" revision="0ec7c65fa6f7592492c4ae73536408caf25c29b3"
Fix reliance on using `str` for extracting an element's CDATA.

Due to an unfortunate design choice, the typical interface to extract
the CDATA inside a `domish.Element` is using `str` on the element and is
implemented by `__str__` returning the first string-like object in its
list of children, regardless of being bytes or unicode strings.
See #3579 for details.

For now, use `unicode` instead, even though the actual implementation is
in `__str__`. This bug became more obvious because of the change to
`io.StringIO` for the Python 3 port.
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_16"></a>|@ralphm commented|
|-|-|

In changeset a52dc78cadcd0633ff07ff93eeacc014d0fd4ca1
```
#!CommitTicketReference repository="" revision="a52dc78cadcd0633ff07ff93eeacc014d0fd4ca1"
Make bytes() return the UTF-8, unicode() the unicode of the CDATA.

This addresses #3579.
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_17"></a>|@ralphm set status to `closed`|
|-|-|

In changeset 43721efe0b48729a8c2c1335947c6ad7981f0426
```
#!CommitTicketReference repository="" revision="43721efe0b48729a8c2c1335947c6ad7981f0426"
Merge py3-xish-8337: Port twisted.words.xish to Python 3.

Author: ralphm
Reviewer: adiroiban
Fixes: #8337, #3579
```
